### PR TITLE
Fix performance regression caused by local tunnel v2

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number:  ref #7181

Problem Summary:

Performance of TPCH100 has regression after local tunnel v2 is merged.

In v2, compared to reactor, local tunnel has too many threads to push data which makes reactor hard to get the chance of pushing data.

### What is changed and how it works?

Add lock to control the concurreny of local tunnel.

  | V1 | V2
-- | -- | --
Q1 | 2.92 | 2.85
Q2 | 1.17 | 1.17
Q3 | 3.12 | 2.72
Q4 | 2.45 | 2.45
Q5 | 6.48 | 6.21
Q6 | 0.64 | 0.64
Q7 | 2.92 | 2.85
Q8 | 5.07 | 4.26
Q9 | 17.48 | 17.35
Q10 | 3.72 | 3.79
Q11 | 0.84 | 0.84
Q12 | 1.51 | 1.58
Q13 | 3.12 | 3.05
Q14 | 0.77 | 0.77
Q15 | 1.64 | 1.64
Q16 | 0.91 | 0.84
Q17 | 6.34 | 6.14
Q18 | 7.08 | 6.95
Q19 | 1.78 | 1.78
Q20 | 1.44 | 1.51
Q21 | 8.82 | 8.89
Q22 | 0.7 | 0.77
Sum | 80.92 | 79.05

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
please do not add internal link in GitHub.

File: dbms/src/Flash/Mpp/MPPTunnel.h [hunk 1]
```diff
@@ -348,6 +334,30 @@ class LocalTunnelSenderV2 : public TunnelSender
...
+        // grpc thread is hard to get chance to push data into MPMCQueue in ExchangeReceiver.
+        // So, we need to control the concurrency of local tunnel threads by the lock.
+        if constexpr (local_only)
+            return local_request_handler.write<enable_fine_grained_shuffle, non_blocking>(source_index, data);
+        else
+        {
+            std::lock_guard lock(mu);
+            return local_request_handler.write<enable_fine_grained_shuffle, non_blocking>(source_index, data);
+        }
+    }
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPTunnel.h:359]: also for `isReadyForWrite`

File: dbms/src/Flash/Mpp/MPPTunnel.h [hunk 2]
```diff
@@ -348,6 +334,30 @@ class LocalTunnelSenderV2 : public TunnelSender
...
+
+        // receiver_mem_tracker pointer will always be valid because ExchangeReceiverBase won't be destructed
+        // before all local tunnels are destructed so that the MPPTask which contains ExchangeReceiverBase and
+        // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
+        data->switchMemTracker(local_request_handler.recv_mem_tracker);
+
+        // When ExchangeReceiver receives data from local and remote tiflash, number of local tunnel threads
+        // is very large and causes the time of transfering data by grpc threads becomes longer, because
+        // grpc thread is hard to get chance to push data into MPMCQueue in ExchangeReceiver.
+        // So, we need to control the concurrency of local tunnel threads by the lock.
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPTunnel.h:351]: Better to add more comments.
```suggestion
        // Adding a lock ensures that there is only one other thread competing with async reactor, so the probability of async reactor getting the lock is 1/2.
        // If the number of local tunnel sender threads is n, then the probability of a local tunnel sender thread getting the lock 
decreases from 1/(1+n) to 1/2 * 1/n = 1/2n
```

**Review by solotzg — COMMENTED** [review]
lgtm

File: dbms/src/Flash/Mpp/MPPTunnel.h (hunk 1 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/Mpp/MPPTunnel.h:359]: > also for `isReadyForWrite`

okk

File: dbms/src/Flash/Mpp/MPPTunnel.h (hunk 2 above)
Comment (xzhangxian1008) [reply dbms/src/Flash/Mpp/MPPTunnel.h:351]: > Better to add more comments.

okk

**Review by SeaRise — APPROVED** [review]
lgtm

**Review by gengliqi — APPROVED** [review]

**Pull request merged**